### PR TITLE
Active Container

### PR DESCRIPTION
**Pull request by DeepSOIC**
Introduces:
* Container interface object
* App.ActiveContainer

Some trivial ports of PartDesign done to make PartDesign usable.

See full description in:
https://forum.freecadweb.org/viewtopic.php?f=20&t=22077

### CODE REVIEW

**Comment by DeepSOIC** [comment]
Travis build failed for Apple+Qt4. Looks unrelated to this PR. Fails on master too.

File: src/Mod/Sketcher/Gui/Command.cpp [hunk 1]
```diff
@@ -215,13 +215,6 @@ void CmdSketcherNewSketch::activated(int iMsg)
         doCommand(Gui,"Gui.activeDocument().setEdit('%s')",FeatName.c_str());
 
         Part::Feature *part = static_cast<Part::Feature*>(support.getValue());//if multi-part support, this will return 0
```
Comment (wwmayer) [inline src/Mod/Sketcher/Gui/Command.cpp]: Leaving this line doesn't make sense, should be removed too.

**Comment by wwmayer** [comment]
Is this PR only for testing on the various platforms or is supposed to be merged to master? For the latter I am still missing unit tests. Since this PR changes a lot of core functionality I have to insist on unit tests for the container stuff.

**Comment by DeepSOIC** [comment]
> Since this PR changes a lot of core functionality I have to insist on unit tests for the container stuff.

OK, will do soon.

**Comment by bblacey** [comment]
Also, it looks like the macOS build failed due to an update to the Eigen port that conflicts with the Orocos KDL Eigen 3.3.2 requirement.  I am in the midst of some busy travel with my family but I will see if I can provide a simple fix.  I do not have time for extensive testing but worst-case if the fix doesn't work, the macOS builds will just remain broken until I return home in a couple weeks or someone else (@ianreese or @peterl94) steps in.

**Comment by DeepSOIC** [comment]
updated!
* a couple of tweaks to container code
* added unit tests
* removed the line of code in sketcher as pointed out by Werner

**Comment by wwmayer** [comment]
Sorry for being annoying. On travis clang and gcc complain about a few more warnings. Here are the files and line numbers. For more information have a look at the build logs:
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/ContainerBasePyImp.cpp:253:60: warning: unused parameter 'attr' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:362:45: warning: unused parameter 'self' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:362:77: warning: unused parameter 'kwd' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:375:51: warning: unused parameter 'self' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:375:83: warning: unused parameter 'kwd' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:393:49: warning: unused parameter 'self' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:393:81: warning: unused parameter 'kwd' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:415:48: warning: unused parameter 'self' [-Wunused-parameter]
/home/travis/build/FreeCAD/FreeCAD/src/App/Containers/Container.cpp:415:80: warning: unused parameter 'kwd' [-Wunused-parameter]

File: src/App/Containers/ContainerBasePyImp.cpp [hunk 2]
```diff
@@ -157,6 +157,10 @@ PyObject* ContainerBasePy::getObject(PyObject* args)
 PyObject* ContainerBasePy::hasObject(PyObject* args)
 {
     PyObject* obj = nullptr;
+    if (PyArg_ParseTuple(args, "O!", &(DocumentPy::Type), &obj))
```
Comment (wwmayer) [inline src/App/Containers/ContainerBasePyImp.cpp]: Is it the desired behaviour that if the argument is a Document it returns False? At least adding a comment would be good to make this clear.

**Comment by DeepSOIC** [comment]
Done! Sorry about me falling into the same "trap" again.

File: src/App/Containers/ContainerPy.xml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      TwinPointer="Container" 
+      Include="App/Containers/Container.h" 
+      FatherInclude="App/Containers/ContainerBasePy.h" 
+      Namespace="App" 
+      Constructor="true"
+      Delete="true"
+      FatherNamespace="App">
+    <Documentation>
+      <Author Licence="LGPL" Name="DeepSOIC" EMail="vv.titov@gmail.com" />
+      <DeveloperDocu>Container - object container interace</DeveloperDocu>
```
Comment (abdullahtahiriyo) [inline src/App/Containers/ContainerPy.xml]: @DeepSOIC 

interace -> interface

**Comment by wwmayer** [comment]
A rebased branch of this PR is available at https://github.com/wwmayer/FreeCAD/tree/ActiveContainer_rebased
See also the forum thread for further discussions.

**Comment by DeepSOIC** [comment]
PR closed for rework.

**Pull request closed without merging**